keps/sig-scheduling/2926-job-mutable-scheduling-directives/README.md

```diff
@@ -0,0 +1,876 @@
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Implement a queue controller.
+- Allow mutating scheduling bits of jobs that started at least once. This will not 
+  allow queue controllers to support preempting jobs and schedule them on a different 
+  partition of the cluster. As discussed in alternatives, this can be addressed in the future.
+- Allow mutating scheduling bits of pods.
+- Allow mutating scheduling bits of apps other than jobs.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+The proposal is to relax update validation of scheduling bits of jobs that have never
+been unsuspended, specifically node affinity, node selector and tolerations.
+
+This has no impact on the job-controller. The job controller has no dependency
+on those scheduling directives expressed in a job's pod template.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to build a controller that implements job queueing and influences when and where a 
+job should run. Users create v1.Job objects, and to control when the job can run, I have 
+a webhook that forces the jobs to be created in a suspended state, and the controller unsuspends 
+them when capacity becomes available.
+
+At the time of creating the job, we may not know on which part of the cluster (.e.g., a zone or 
+a VM type) a job should run. To control where a job should run, the controller can update 
+node affinity or tolerations of the Job's pod template; by doing so, the queue controller is 
+able to influence the scheduler and autoscaler decisions.
+
+<!--
+#### Story 2
+
+
+### Notes/Constraints/Caveats (Optional)
+-->
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- New calls from a queue controller to the API server to update affinity or tolerations. 
+  The mitigation is for such controllers to make a single API call for both updating
+  affinity/tolerations and unsuspending the job.
+  
+- A race condition could happen that may result in having a mix of pods with different 
+  affinity/tolerations, consider the following scenario: 
+  1. unsuspend the job
+  2. suspend the job
+  3. update affinity
+  
+  After unsuspending the job, the job-controller could have created some pods and not yet
+  updated `Status.StartTime` by the time kube-apiserver received the suspend then affinity updates.
```

Comment:
I actually thought that Status.StartTime is set atomically with unsuspending (by kube-apiserver) and there isn't a race.
So we effectively switched one race to a different one.

I'm not a fan of it, but I can probably live with it. But I would also like to see an ACK from @liggitt [although knowing Jordan, he was aware of it].

Reply:
So StartTime is set by the job-controller. I mentioned in https://github.com/kubernetes/kubernetes/issues/104714#issuecomment-914891648 that there is a race, but I feel this is a lot less common than the one where we allow it for any suspended job and it is something we can live with until we refactor the job controller.

Reply:
that seems ok... that's a much smaller / easier to reason about window. the race could even be completely closed by making the job controller do a write to set startTime and then wait for the resulting update event to resync / start spawning pods